tripal/tripal
# Update ChadoProperty Text widget to support filtered HTML

# Tripal 4 Core Dev Task

### Issue #1974 


### Tripal Version: 4

## Description


With PR #1862 we added support for the WYSIWYG toolbar for text fields, and now we need to do the same thing for property fields. Furthermore, it would be good to allow admin to configure the number of rows for the textarea.

This is needed because properties are often used to store metadata which can be more then a single value. For example, ActivatingCrops uses properties to store methodology and experimental design for research studies. This are often long text and sometimes even include links and tables.

Additionally, ensure that all the other property widgets return filtered text because on 4.x curators can enter HTML into the property field even if they do not have the appropriate permissions.

Specifically, this PR
 - adds a filter setting to the "Chado Property: Long Text" widget that defaults to basic.
 - ensures that the Chado Property formatter uses this setting.
 - since all Chado Property widgets share the same formatter, ensure that basic_html is used for all other widgets or if the "Chado Property: Long Text" was not configured.
 - ensures the property widgets respect the "Required" setting set in the field settings.
 - allows admin to choose the number of rows shown in the "Chado Property: Long Text" widget

## Testing?



### Manual

1. Add 4 different property fields to the organism content type. The terms do not matter. Set all to unlimited values.
2. Configure the form display and set two of these fields to use the long text widget and the other two to use the other widgets. Configure the long text widgets to use "Basic" for one and "Full" for the other. Add some options to the select list widget.
3. Create content. Enter the same block of HTML into both long text widgets. Ensure this includes an HTML. Also enter values in the other two widgets.
4. Save and confirm that the long text with "Basic" setting has stripped out the table/list formatting but the one using "Full" has not. We have two fields here to also confirm that the setting is on a per field basis.
5. Check that no errors were thrown for the other two property fields (ie. string and select widget) even though they did not have a filter format configured.

Other things to test:
- set some of the fields to required and others not and confirm the form respects this.
- change the format from "Full" to "Basic" and confirm the formatter/page strips tables without you needing to edit the page. Also check that when you do edit the page, the HTML is stripped.
- change the setting to indicate the number of rows for the "Long Text" widget and confirm the widget reflects this setting.
- enter an HTML table one-liner (e.g. `<table><thead><tr><th>H1</th><th>H2</th></tr></thead><tbody><tr><td>D1</td><td>D2</td></tr></tbody></table>`) as an option for the "Select" widget when configuring it and confirm that while you can select it, when the page is viewed, the HTML is stripped. You should be able to use basic html in a select option (i.e. bold or italics)

### Automated Testing

I started adding automated testing for the property field based on the model I used for Tripal fields lately. My progress is included here and, while not complete, it is fully functioning and better coverage then we had previously.

Includes
 - ChadoFieldTestTrait: this is the Chado-focused version of the TripalFieldTestTrait. It includes methods for creating field types + instances as well as an assert to confirm the values match what was expected. It also includes a method to make setting up the test environment for field kernel tests easier.
 - ChadoPropertyTypeSimpleCRUDTest: focuses on testing a single property field on a project-based content type with only one delta. It provides 44% test coverage of the ChadoPropertyTypeDefault field type class and tests creating a TripalEntity with a property field attached. Specifically, it checks that
   - the property field can be created programatically and attached to a tripal entity type.
   - a tripal entity can be created programatically with values for a property field supplied.
   - the property field values are set properly in the tripal entity class
   - tripal entity can retrieve the values successfully which checks basic interactions between the fields and the TripalStorage backend
   - that the entity can be saved without throwing any errors
   - that tripalTypes can be called without error and all the property type objects are created as expected.

Test coverage shows that this test is travelling the expected paths through the ChadoPropertyTypeDefault, TripalFieldItemBase, TripalEntity, and ChadoStorage for inserting field data.

Next steps are to check the field values are inserted into chado; however, this is not producing the expected results... Since I have already spent more time then I should have, I am moving on. These tests are included because they provide a good model for moving forward on testing fields.
